Fixes memory leak on JDK 16+

caesar-ralf opened the pull request:
In newer versions of java, specifically 16 and over, the `ThreadLocal` can
be cleared before the Thread is killed. This breaks the assumption of
`ThreadLocalHolder` which expected the `initializeThread` method to be
executed only once per `Thread`.

This commit changes the `threadValues` to become a map that holds which
thread the value is associated with so we can avoid recreating new
values when one already exists.

Fixes #722

caesar-ralf commented:
Ping @brenuart, could you take a look at this fix for the memory leak? I am sorry to ping you directly, but it's been a week and we are blocked off because of this problem.

brenuart commented on src/main/java/net/logstash/logback/util/ThreadLocalHolder.java:
```diff
@@ -148,39 +149,42 @@ public void close() {
-        return holder;
+        final Thread currentThread = Thread.currentThread();
+        final long threadId = currentThread.getId();
+        return threadValues.computeIfAbsent(
```
I would add a comment here about the reason why the value associated with the current thread is stored in the threadValues Map and not in the ThreadLocal itself as it was before. Making reference to the ForkJoinPool issue clearing the thread local while the thread is still alive would be nice as well.

Note that this strategy makes the assumption that Thread.getId() does not change after a thread is created. This seems to be confirmed by the javadoc: the id seems to be immutable and unique...

brenuart replied:
Also, you need to rebase to pickup the latest changes otherwise the build will keep failing... (not related to your PR)

caesar-ralf replied:
added

brenuart reviewed (commented):
Looks nice. Relying on the Thread#getId and a Map to store the values instead of the ThreadLocal seems to be the only option.

One small general remark: your PR also contains cosmetic changes like removal of blanks... these changes produce some noise and make it harder to spot what is important :(

Thanks for your work. I'll merge it asap once you reviewed my other comments.

brenuart commented:
Also, you need to rebase to pickup the latest changes otherwise the build will keep failing... (not related to your PR)

caesar-ralf commented:
> One small general remark: your PR also contains cosmetic changes like removal of blanks... these changes produce some noise and make it harder to spot what is important :(

ow yeah, sorry for that. That was my IDE doing automatically. I will work on fixing the issues raised by the PR

caesar-ralf commented:
> Also, you need to rebase to pickup the latest changes otherwise the build will keep failing... (not related to your PR)

Changed it. I checked why it was failing in jdk8 and it seemed to be a hiccup because it couldn't download the javadocs. Feel free to change the comment I added as you please

brenuart commented:
Merged. Thanks for your work.

caesar-ralf commented:
Thanks for helping out @brenuart and being understanding! Can you let me know when you do release with this fix?

brenuart commented:
Soon (hopefully by the end of the week). Just waiting for another PR to be merged and we are all set ;-)

> Le 19 mai 2022 à 13:59, Caesar Ralf Franz Hoppen ***@***.***> a écrit :
> 
> ﻿
> Thanks for helping out @brenuart and being understanding! Can you let me know when you do release with this fix?
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

fredericdasilva commented:
Thanks a lot. This commit fixed memory leaks that we faced in Production since last upgrade on JDK 17.
